mujoco_warp/_src/sensor.py

```diff
@@ -1860,6 +1862,134 @@ def _sensor_touch(
+  body_weldid: wp.array(dtype=int),
+  geom_bodyid: wp.array(dtype=int),
+  mesh_vertadr: wp.array(dtype=int),
+  mesh_vert: wp.array(dtype=wp.vec3),
+  mesh_normaladr: wp.array(dtype=int),
+  mesh_normal: wp.array(dtype=wp.vec3),
+  mesh_quat: wp.array(dtype=wp.quat),
+  sensor_objid: wp.array(dtype=int),
+  sensor_refid: wp.array(dtype=int),
+  sensor_dim: wp.array(dtype=int),
+  sensor_adr: wp.array(dtype=int),
+  plugin: wp.array(dtype=int),
+  plugin_attr: wp.array(dtype=wp.vec3f),
+  geom_plugin_index: wp.array(dtype=int),
+  taxel_vertadr: wp.array(dtype=int),
+  taxel_sensorid: wp.array(dtype=int),
+  # Data in:
+  ncon_in: wp.array(dtype=int),
+  geom_xpos_in: wp.array2d(dtype=wp.vec3),
+  geom_xmat_in: wp.array2d(dtype=wp.mat33),
+  subtree_com_in: wp.array2d(dtype=wp.vec3),
+  cvel_in: wp.array2d(dtype=wp.spatial_vector),
+  contact_geom_in: wp.array(dtype=wp.vec2i),
+  contact_worldid_in: wp.array(dtype=int),
+  # Data out:
+  sensordata_out: wp.array2d(dtype=float),
+):
+  conid, taxelid = wp.tid()
+
+  if conid >= ncon_in[0]:
+    return
+
+  worldid = contact_worldid_in[conid]
+
+  # get sensor_id
+  sensor_id = taxel_sensorid[taxelid]
+
+  # get parent weld id
+  mesh_id = sensor_objid[sensor_id]
+  geom_id = sensor_refid[sensor_id]
+  parent_body = geom_bodyid[geom_id]
+  parent_weld = body_weldid[parent_body]
+
+  # contact geom
+  body1 = body_weldid[geom_bodyid[contact_geom_in[conid][0]]]
+  body2 = body_weldid[geom_bodyid[contact_geom_in[conid][1]]]
+  if body1 == parent_weld:
+    geom = contact_geom_in[conid][1]
+  elif body2 == parent_weld:
+    geom = contact_geom_in[conid][0]
+  else:
+    return
+  body = geom_bodyid[geom]
+
+  # vertex local position
+  vertid = taxel_vertadr[taxelid] - mesh_vertadr[mesh_id]
+  pos = mesh_vert[vertid + mesh_vertadr[mesh_id]]
+
+  # position in global frame
+  xpos = geom_xmat_in[worldid, geom_id] @ pos
+  xpos += geom_xpos_in[worldid, geom_id]
+
+  # position in other geom frame
+  tmp = xpos - geom_xpos_in[worldid, geom]
+  lpos = wp.transpose(geom_xmat_in[worldid, geom]) @ tmp
+
+  # compute distance
+  plugin_id = geom_plugin_index[geom]
+  depth = wp.min(sdf(int(GeomType.SDF.value), lpos, plugin_attr[plugin_id], plugin[plugin_id]), 0.0)
+  if depth >= 0.0:
+    return
+
+  # get velocity in global
+  vel_sensor = _transform_spatial(cvel_in[worldid, parent_weld], xpos - subtree_com_in[worldid, body_rootid[parent_weld]])
+  vel_other = _transform_spatial(
+    cvel_in[worldid, body], geom_xpos_in[worldid, geom] - subtree_com_in[worldid, body_rootid[body]]
+  )
+  vel_rel = vel_sensor - vel_other
+
+  # get contact force/torque, rotate into node frame
+  offset = mesh_normaladr[mesh_id] + 3 * vertid
+  normal = math.rot_vec_quat(mesh_normal[offset], mesh_quat[mesh_id])
+  tang1 = math.rot_vec_quat(mesh_normal[offset + 1], mesh_quat[mesh_id])
+  tang2 = math.rot_vec_quat(mesh_normal[offset + 2], mesh_quat[mesh_id])
+  kMaxDepth = 0.05
+  pressure = depth / (kMaxDepth - depth)
```

Comment:
can there be division by zero here?

Reply:
Fixed by also preventing negative values.